File: instagram_client.py
import hashlib
import random
import re
import time
from typing import Optional, Tuple
from playwright.sync_api import BrowserContext, Page

from config import (
    HEADLESS,
    STORY_DELAY_MIN,
    STORY_DELAY_MAX,
    POST_DELAY_MIN,
    POST_DELAY_MAX,
)
from parsers import normalize_instagram_url


def is_instagram_logged_in(context: BrowserContext) -> bool:
    """Checks if the user has an active Instagram sessionid cookie."""
    for cookie in context.cookies():
        if cookie.get("name") == "sessionid" and cookie.get("value"):
            return True
    return False


def wait_for_login(context: BrowserContext, page: Page) -> bool:
    """
    Ensures Instagram session is authenticated.
    Prompts user to log in via browser window if running interactively.
    """
    if is_instagram_logged_in(context):
        print("✅ Active Instagram session found.")
        return True

    if HEADLESS:
        print("\n❌ No active session found and running with HEADLESS=True.")
        print("Please run once with HEADLESS=0 to log in interactively:")
        print("HEADLESS=0 python3 story_monitor.py")
        return False

    print("\n==============================")
    print("INSTAGRAM LOGIN")
    print("==============================")

    page.goto(
        "https://www.instagram.com/accounts/login/",
        wait_until="domcontentloaded",
        timeout=60_000,
    )

    print("\nPlease log in to your Instagram account in the opened browser window.")
    print("The script will automatically detect the login and proceed...")

    max_wait_seconds = 600
    waited = 0

    while waited < max_wait_seconds:
        if is_instagram_logged_in(context):
            print("\n✅ Instagram login successful!")
            page.wait_for_timeout(2000)
            return True

        if waited % 10 == 0 and waited > 0:
            print(f"Waiting for login... ({waited}/{max_wait_seconds}s)")

        time.sleep(2)
        waited += 2

    print("\n❌ Login timed out.")
    return False


def get_story_fingerprint(page: Page) -> str:
    """Computes a unique SHA-256 fingerprint for the current story view."""
    try:
        text = page.locator("body").inner_text(timeout=2000)
    except Exception:
        text = ""

    try:
        media = page.locator("img, video").evaluate_all(
            """
            elements => elements.map(el => {
                const r = el.getBoundingClientRect();
                return [el.currentSrc || el.src || el.poster || '', r.width, r.height].join(':');
            }).join('|')
            """
        )
    except Exception:
        media = ""

    raw = f"{page.url}|{text[:3000]}|{media[:3000]}"
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()


def find_post_in_story(page: Page) -> Optional[str]:
    """Finds shared /p/ or /reel/ links inside the current story."""
    found = []

    for selector in ['a[href*="/p/"]', 'a[href*="/reel/"]']:
        locator = page.locator(selector)
        try:
            count = locator.count()
        except Exception:
            continue

        for i in range(count):
            try:
                href = locator.nth(i).get_attribute("href")
            except Exception:
                continue

            url = normalize_instagram_url(href)
            if url and ("/p/" in url or "/reel/" in url):
                found.append(url)

    # Return first unique post URL
    unique = list(dict.fromkeys(found))
    return unique[0] if unique else None


def go_to_next_story(page: Page) -> bool:
    """Navigates to the next story slide with randomized anti-bot delay."""
    before = get_story_fingerprint(page)

    # Human-like random delay before next slide
    delay = random.uniform(STORY_DELAY_MIN, STORY_DELAY_MAX)
    time.sleep(delay)

    try:
        page.keyboard.press("ArrowRight")
        page.wait_for_timeout(1000)
    except Exception:
        pass

    after = get_story_fingerprint(page)
    if before != after:
        return True

    # Fallback to mouse click on right edge of story viewport
    try:
        viewport = page.viewport_size
        if viewport:
            page.mouse.click(
                viewport["width"] - 120,
                viewport["height"] // 2,
            )
            page.wait_for_timeout(1000)
    except Exception:
        pass

    return before != get_story_fingerprint(page)


def clean_raw_caption(raw: str) -> str:
    """Strips likes, comments, and date wrappers from Instagram captions."""
    if not raw:
        return ""

    # Strip og:description wrapper: 'XX likes, YY comments - username on Date: "..."'
    cleaned = re.sub(
        r"^\d+[\s\w,]*likes?,[\s\w,]*comments?\s*[-–—]\s*[\w.]+(?:\s+on\s+[^:]+)?:\s*[\"“']?",
        "",
        raw.strip(),
        flags=re.IGNORECASE,
    )
    # Strip trailing quotes
    cleaned = re.sub(r"[\"”']+$", "", cleaned.strip())
    return cleaned.strip()


def get_post_info(context: BrowserContext, post_url: str) -> Tuple[str, Optional[str]]:
    """
    Opens the shared post/reel in a separate tab with anti-bot delay to extract clean caption and cover image,
    then closes the tab without disrupting the active story loop.
    """
    post_page = context.new_page()

    try:
        post_page.goto(
            post_url,
            wait_until="domcontentloaded",
            timeout=45_000,
        )

        # Human-like delay while post renders
        delay = random.uniform(POST_DELAY_MIN, POST_DELAY_MAX)
        time.sleep(delay)

        caption = ""
        image_url = None

        # 1. Caption extraction from article
        article = post_page.locator("article")
        if article.count() > 0:
            try:
                caption = article.first.inner_text(timeout=5000).strip()
            except Exception:
                pass

        # 2. Fallback to meta og:description
        if not caption:
            og_desc = post_page.locator('meta[property="og:description"]')
            if og_desc.count() > 0:
                caption = (og_desc.first.get_attribute("content") or "").strip()

        # Clean any likes/comments wrapper noise
        caption = clean_raw_caption(caption)

        # 3. Image extraction (og:image or largest article image)
        og_image = post_page.locator('meta[property="og:image"]')
        if og_image.count() > 0:
            image_url = og_image.first.get_attribute("content")

        if not image_url:
            try:
                image_url = post_page.locator("article img").evaluate_all(
                    """
                    elements => elements.map(el => {
                        const r = el.getBoundingClientRect();
                        return {
                            url: el.currentSrc || el.src || '',
                            area: r.width * r.height
                        };
                    }).filter(x => x.url && x.area > 30000)
                      .sort((a, b) => b.area - a.area)[0]?.url || null
                    """
                )
            except Exception:
                pass

        return caption, image_url

    finally:
        post_page.close()

File: src/instagramClient.ts
import crypto from 'node:crypto';
import type { BrowserContext, Page } from 'playwright';
import {
  HEADLESS,
  STORY_DELAY_MIN,
  STORY_DELAY_MAX,
  POST_DELAY_MIN,
  POST_DELAY_MAX,
} from './config.js';
import { normalizeInstagramUrl, cleanCaptionText } from './parsers.js';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function randomBetween(min: number, max: number): number {
  return Math.random() * (max - min) + min;
}

export async function isInstagramLoggedIn(context: BrowserContext): Promise<boolean> {
  const cookies = await context.cookies();
  return cookies.some((c) => c.name === 'sessionid' && Boolean(c.value));
}

export async function waitForLogin(context: BrowserContext, page: Page): Promise<boolean> {
  const loggedIn = await isInstagramLoggedIn(context);
  if (loggedIn) {
    console.log('✅ Active Instagram session found.');
    return true;
  }

  if (HEADLESS) {
    console.log('\n❌ No active session found and running with HEADLESS=true.');
    console.log('Please run once with HEADLESS=0 to log in interactively:');
    console.log('HEADLESS=0 npm start');
    return false;
  }

  console.log('\n==============================');
  console.log('INSTAGRAM LOGIN');
  console.log('==============================');

  await page.goto('https://www.instagram.com/accounts/login/', {
    waitUntil: 'domcontentloaded',
    timeout: 60000,
  });

  console.log('\nPlease log in to your Instagram account in the opened browser window.');
  console.log('The script will automatically detect the login and proceed...');

  const maxWaitSeconds = 600;
  let waited = 0;

  while (waited < maxWaitSeconds) {
    if (await isInstagramLoggedIn(context)) {
      console.log('\n✅ Instagram login successful!');
      await page.waitForTimeout(2000);
      return true;
    }

    if (waited % 10 === 0 && waited > 0) {
      console.log(`Waiting for login... (${waited}/${maxWaitSeconds}s)`);
    }

    await sleep(2000);
    waited += 2;
  }

  console.log('\n❌ Login timed out.');
  return false;
}

export async function getStoryFingerprint(page: Page): Promise<string> {
  let text = '';
  try {
    text = await page.locator('body').innerText({ timeout: 2000 });
  } catch {}

  let media = '';
  try {
    const list = await page.locator('img, video').evaluateAll((elements) =>
      elements.map((el) => {
        const r = el.getBoundingClientRect();
        const src =
          (el as HTMLImageElement).currentSrc ||
          (el as HTMLImageElement).src ||
          (el as HTMLVideoElement).poster ||
          '';
        return `${src}:${r.width}:${r.height}`;
      })
    );
    media = list.join('|');
  } catch {}

  const raw = `${page.url()}|${text.slice(0, 3000)}|${media.slice(0, 3000)}`;
  return crypto.createHash('sha256').update(raw).digest('hex');
}

export async function findPostInStory(page: Page): Promise<string | null> {
  const found: string[] = [];

  for (const selector of ['a[href*="/p/"]', 'a[href*="/reel/"]']) {
    const locator = page.locator(selector);
    let count = 0;
    try {
      count = await locator.count();
    } catch {
      continue;
    }

    for (let i = 0; i < count; i++) {
      try {
        const href = await locator.nth(i).getAttribute('href');
        const url = normalizeInstagramUrl(href);
        if (url && (url.includes('/p/') || url.includes('/reel/'))) {
          found.push(url);
        }
      } catch {}
    }
  }

  const unique = Array.from(new Set(found));
  return unique[0] || null;
}

export async function goToNextStory(page: Page): Promise<boolean> {
  const before = await getStoryFingerprint(page);

  // Human-like random delay
  const delay = randomBetween(STORY_DELAY_MIN, STORY_DELAY_MAX);
  await sleep(delay * 1000);

  try {
    await page.keyboard.press('ArrowRight');
    await page.waitForTimeout(1000);
  } catch {}

  const after = await getStoryFingerprint(page);
  if (before !== after) return true;

  // Fallback to mouse click on right side of viewport
  try {
    const viewport = page.viewportSize();
    if (viewport) {
      await page.mouse.click(viewport.width - 120, Math.floor(viewport.height / 2));
      await page.waitForTimeout(1000);
    }
  } catch {}

  const finalCheck = await getStoryFingerprint(page);
  return before !== finalCheck;
}

export async function getPostInfo(
  context: BrowserContext,
  postUrl: string
): Promise<[string, string | null]> {
  const postPage = await context.newPage();

  try {
    await postPage.goto(postUrl, {
      waitUntil: 'domcontentloaded',
      timeout: 45000,
    });

    const delay = randomBetween(POST_DELAY_MIN, POST_DELAY_MAX);
    await sleep(delay * 1000);

    let caption = '';
    let imageUrl: string | null = null;

    // 1. Extract from article
    const article = postPage.locator('article');
    if ((await article.count()) > 0) {
      try {
        caption = (await article.first().innerText({ timeout: 5000 })).trim();
      } catch {}
    }

    // 2. Fallback to meta og:description
    if (!caption) {
      const ogDesc = postPage.locator('meta[property="og:description"]');
      if ((await ogDesc.count()) > 0) {
        caption = ((await ogDesc.first().getAttribute('content')) || '').trim();
      }
    }

    // Clean caption wrapper noise
    caption = cleanCaptionText(caption);

    // 3. Image extraction (og:image or largest article image)
    const ogImage = postPage.locator('meta[property="og:image"]');
    if ((await ogImage.count()) > 0) {
      imageUrl = await ogImage.first().getAttribute('content');
    }

    if (!imageUrl) {
      try {
        imageUrl = await postPage.locator('article img').evaluateAll((elements) => {
          const candidates = elements
            .map((el) => {
              const r = el.getBoundingClientRect();
              return {
                url: (el as HTMLImageElement).currentSrc || (el as HTMLImageElement).src || '',
                area: r.width * r.height,
              };
            })
            .filter((x) => x.url && x.area > 30000)
            .sort((a, b) => b.area - a.area);

          return candidates[0]?.url || null;
        });
      } catch {}
    }

    return [caption, imageUrl];
  } finally {
    await postPage.close();
  }
}
